manta-pay/src/crypto/hash/poseidon_parameter_generation/constants.rs

```diff
@@ -0,0 +1,109 @@
+// Copyright 2019-2022 Manta Network.
+// This file is part of manta-rs.
+//
+// manta-rs is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// manta-rs is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with manta-rs.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Generate constants for optimized poseidon hash
+
+use super::round_numbers::calc_round_numbers;
+use super::round_constants::generate_round_constants;
+use super::preprocessing::compress_round_constants;
+use super::matrix::Matrix;
+use super::mds::{factor_to_sparse_matrixes, MdsMatrices, SparseMatrix};
+use crate::crypto::hash::poseidon::Specification;
+use core::fmt::Debug;
+use alloc::vec::Vec;
+
+// TODO: shall we put constant generation code to compile time?
+
+#[derive(Clone)]
+/// Poseidon Constants struct for optimized poseidon
+pub struct PoseidonConstants<S, COM>
+where 
+S: Specification<COM> + Clone,
+S::ParameterField: Clone,
+{
+    /// mds matrix data
+    pub mds_matrices: MdsMatrices<S, COM>,
+    /// a vector of round constants
+    pub round_constants: Vec<S::ParameterField>,
+    /// The following three operators are used for optimized poseidon hash
+    pub compressed_round_constants: Vec<S::ParameterField>,
+    pub pre_sparse_matrix: Matrix<S, COM>,
+    pub sparse_matrixes: Vec<SparseMatrix<S, COM>>,
+    /// domain tag for domain separation
+    pub domain_tag: S::ParameterField,
+    /// number of full rounds. Note: full_rounds % 2 == 0
+    pub full_rounds: usize,

```

Comment:
We might want to hardcode `full_rounds`, `half_full_rounds`, and `partial_rounds` as a spec (e.g. associated constant of a trait) instead of calculating on runtime. Check @bhgomes's original implementation for an example.

Reply:
This requires some nontrivial work and careful design. We can do this later if you want, or I can handle this when I have time.

Reply:
This part is only used in the optimized Poseidon hash. Considering 1) this PR is already kind of complex, and 2) the optimized Poseidon hash does not bring benefits on Groth16, I would like to suggest keeping this part and update in the next PR when we adapt the optimized Poseidon hash.